doc/development/powershell.md

```diff
@@ -0,0 +1,179 @@
+    - Use the new `$PSNativeCommandUseErrorActionPreference` variable coming in 7.3 (https://docs.microsoft.com/en-us/powershell/scripting/whats-new/what-s-new-in-powershell-73?view=powershell-7.2#experimental-features).
+- Set global powershell variables to achieve consistent behavior:
+  - `$ErrorActionPreference` should be set to `Stop` at the top of any cmdlet files unless there is a specific reason to do something else.
+  - `Set-StrictMode -Version 4` should be set at the top of any cmdlet files unless there is a specific reason to do something else.
+- Support the common `-WhatIf` parameter. See [SupportsShouldProcess docs](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.2#supportsshouldprocess).
+  - Scripts should be able to be run locally, and they should also support a dry run mode via `-WhatIf` so a user can see what actions might be performed that could be destructive.
+  - Most common powershell module functions will support `-WhatIf` natively, but engsys code that does any write operations should also gate this functionality behind this parameter by checking [ShouldProcess](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.2#pscmdletshouldprocess).
+- Log write actions with enough context that someone reading the log statement in a pipeline would be able to know what happened.
+  - For example, `Write-Host "Updating code owner to '$user' in '$ServiceDirectory'"` is a much better log than `Write-Host "Updating codeowner"`.
+
+## Style
+
+Powershell is a very flexible language. In engsys we try to use a smaller subset of what's possible to keep programs consistent and their behavior predictable.
+
+- Declare types for function parameters (e.g. `function foo([string]$bar) {}`.
+- Add formatted documentation for entrypoints like top-level scripts or publicly exposed functions.
+  ```
+  <#
+  .SYNOPSIS
+  <blah>
+  
+  .DESCRIPTION
+  <blah>
+  
+  .PARAMETER baz
+  <blah>
+  #>
+  function foobar([string]$baz)
+  ```
+- Avoid complex pipeline statements.
+- Put curly braces `{` for opening function blocks on a new line.
+    ```
+    function foo()
+    {
+    }
+    ```
+- Use explicit `return` statements for returning function/script block values.
+- Favor simple variable substitution in strings (see [variable substitution docs](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-string-substitutions?view=powershell-7.2)).
+  - For simple variables in a string, use syntax like `"my name is $username"`. Do not use `"my name is $($username)"`.
+  - For properties, use syntax like `"my name is $($user.name)"`.
+  - Only use curly braces when the delineation is unclear, e.g. `"Updating all ${itemType}s"`
+- Use full CmdLet names for better scripts.
+  - Favor `$foobar | Where-Object {} | ForEach-Object {}` over `$foobar | ? {} | % {}`.
```

Comment:
I know you are demonstrating a concept here but I also like using .Where({}) over the Where-Object to help avoid the pipeline troubles.

Reply:
I know I posted some guidance on this in EngSys channel but teams search is letting me down. But I do think we should add some more warnings and recommendations to use `.Where(...)` instead.